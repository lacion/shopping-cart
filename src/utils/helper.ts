export const formatError = (nameOfFunction: string, error: Error) => {
  return process.env.NODE_ENV === 'test'
    ? ''
    : console.log(`----${nameOfFunction}----`, error)
}

export const formatPrice = (price: number): string => {
  if (!price) {
    return 'R' + (0).toFixed(2)
  }

  return new Intl.NumberFormat('en-ZA', {
    style: 'currency',
    currency: 'ZAR',
    minimumFractionDigits: 2,
  })
    .format(price / 100)
    .replace(',', '.')
}
