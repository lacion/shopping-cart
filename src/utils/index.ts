export * from './auth'
export * from './helper'
export * from './jwt'
export * from './mocks'
export * from './prisma'
