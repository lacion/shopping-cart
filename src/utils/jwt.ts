import { sign, verify } from 'jsonwebtoken'
import { clone } from 'lodash'
import { Context } from './prisma'

interface Token {
  id: number
}

type User = {
  id: number
}
// get user id from auth token
export function getUserId({ req }: Context): number | undefined {
  if (req) {
    const Authorization = req.headers.authorization || req.headers.Authorization

    if (Authorization) {
      const token = Authorization.replace('Bearer ', '')

      const verifiedToken = verify(
        token,
        process.env.JWT_SECRET as string,
      ) as Token

      return verifiedToken && verifiedToken.id
    }
  }
  return undefined
}

// issue new token based on payload
export const issue = (payload: string | User | Buffer, jwtOptions = {}) => {
  return sign(clone(payload), process.env.JWT_SECRET as string, {
    ...jwtOptions,
    expiresIn: '1d',
  })
}
