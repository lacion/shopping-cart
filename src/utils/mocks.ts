import { Cart, Category, Customer, Product } from '@prisma/client'
import * as Factory from 'factory.ts'
import { address, commerce, datatype, date, internet, name } from 'faker'

export const productFactory = Factory.Sync.makeFactory<Product>({
  id: `${Factory.each((i) => i)}`,
  name: commerce.productName(),
  price: 2500,
  description: commerce.productDescription(),
  expiresAt: date.future(),
  sku: datatype.number(),
  stockLevel: 20,
  categoryId: `${Factory.each((i) => i)}`,
})

export const categoryFactory = Factory.Sync.makeFactory<Category>({
  id: `${Factory.each((i) => i)}`,
  name: 'Food',
})

export const cartFactory = Factory.Sync.makeFactory<Cart>({
  id: `${Factory.each((i) => i)}`,
  customerId: `1`,
  isCheckedOut: false,
  createdAt: Factory.Sync.each(() => new Date()),
  updatedAt: Factory.Sync.each(() => new Date()),
})

export const cartItemFactory = Factory.Sync.makeFactory<Category>({
  id: `${Factory.each((i) => i)}`,
  name: 'Food',
})

export const customerFactory = Factory.Sync.makeFactory<Customer>({
  id: `${Factory.each((i) => i)}`,
  email: internet.email(),
  name: name.findName(),
  address: address.streetAddress(),
})
