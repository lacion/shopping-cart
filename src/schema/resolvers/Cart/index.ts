import { Cart } from '@prisma/client'
import { ApolloError } from 'apollo-server'
import {
  MutationAddToCartArgs,
  MutationCheckoutArgs,
  MutationRemoveFromCartArgs,
  QueryCartArgs,
} from '../../../generated'
import { Context, formatError, formatPrice } from '../../../utils'

export default {
  Query: {
    cart: async (
      _parent: unknown,
      { id }: QueryCartArgs,
      { prisma }: Context,
    ) => {
      try {
        if (!id) {
          throw new ApolloError('Cart ID is required')
        }

        return prisma.cart.findUnique({
          where: {
            id,
          },
        })
      } catch (error) {
        formatError('cart', error)
        return error
      }
    },
  },
  Mutation: {
    addToCart: async (
      _parent: unknown,
      { input }: MutationAddToCartArgs,
      { prisma, userId }: Context,
    ) => {
      try {
        const { productId, quantity, cartId } = input
        let safeQuantity = quantity > 0 ? quantity : 1

        // no specific product id passed.
        if (!productId) {
          throw new ApolloError('Product not found')
        }

        const check = await prisma.product.findFirst({
          where: {
            AND: [{ id: productId }],
          },
        })

        console.log({ check, productId })
        // check if product is in stock based on quantity
        const hasEnoughProducts = await prisma.product.findFirst({
          where: {
            AND: [{ id: productId }, { stockLevel: { gte: safeQuantity } }],
          },
        })

        if (!hasEnoughProducts) {
          throw new ApolloError('Product out of stock')
        }

        // check if user already has unchecked out cart
        let cart: Cart | null

        // if cartId is passed, use that
        if (cartId) {
          console.log('has cartId')
          cart = await prisma.cart.findFirst({
            where: {
              AND: [{ id: cartId }, { isCheckedOut: false }],
            },
          })

          if (!cart) {
            console.log('cart not found/checked out so creating new one')
            // create new cart
            cart = await prisma.cart.create({
              data: {
                customer: { connect: { id: userId } },
              },
            })
          }
        } else {
          // check if user has a cart
          const userCart = await prisma.cart.findFirst({
            where: {
              isCheckedOut: false,
            },
          })

          console.log({ userCart })
          if (userCart) {
            console.log('user has unchecked out cart')
            cart = userCart
          } else {
            console.log('user has no cart so creating new one')
            // create new cart
            cart = await prisma.cart.create({
              data: {
                customer: { connect: { id: userId } },
              },
            })
          }
        }

        console.log({ cart })

        const priceInCents = hasEnoughProducts.price * safeQuantity

        // if this product is already in cart, update cart item quantity & price
        const cartItem = await prisma.cartItem.findFirst({
          where: {
            AND: [{ cart: { id: cart.id } }, { product: { id: productId } }],
          },
        })

        if (cartItem) {
          // add product to cart
          const updatedCartItem = prisma.cartItem.update({
            where: { id: cartItem.id },
            data: {
              quantity: {
                increment: safeQuantity,
              },
              price: {
                increment: priceInCents,
              },
            },
          })

          // update product stock level
          const newStockLevel = prisma.product.update({
            where: { id: productId },
            data: {
              stockLevel: {
                decrement: safeQuantity,
              },
            },
          })

          // wrap operations in transaction to ensure both mutations succeed
          await prisma.$transaction([updatedCartItem, newStockLevel])

          return cart
        }

        if (!cart) {
          throw new ApolloError('Cart not found')
        }

        console.log('create new cart item', cart)
        const reCheck = await prisma.cart.findMany()

        console.log({ reCheck })
        // add product to cart
        const newCartItem = prisma.cartItem.create({
          data: {
            quantity: safeQuantity,
            cart: { connect: { id: cart.id } },
            product: { connect: { id: productId } },
            price: priceInCents,
          },
        })

        // update product stock level
        const newStockLevel = prisma.product.update({
          where: { id: productId },
          data: {
            stockLevel: {
              decrement: safeQuantity,
            },
          },
        })

        // wrap operations in transaction to ensure both mutations succeed
        await prisma.$transaction([newCartItem, newStockLevel])

        return cart
      } catch (error) {
        formatError('addToCart', error)
        return error
      }
    },
    removeFromCart: async (
      _parent: unknown,
      { input }: MutationRemoveFromCartArgs,
      { prisma }: Context,
    ) => {
      try {
        const { cartItemId, cartId } = input

        // no specific item id or cart id passed.
        if (!cartItemId || !cartId) {
          throw new ApolloError('Invalid input')
        }

        // check if cart is valid
        const isValidCart = await prisma.cart.findFirst({
          where: { AND: [{ id: cartId }, { isCheckedOut: false }] },
        })

        if (!isValidCart) {
          throw new ApolloError('Cart not found')
        }

        // check if cart is valid
        const isValidCartItem = await prisma.cartItem.findUnique({
          where: { id: cartItemId },
          include: {
            product: true,
          },
        })

        if (!isValidCartItem) {
          throw new ApolloError('Cart Item not found')
        }

        const quantity = isValidCartItem.quantity

        // update product stock level
        const newStockLevel = prisma.product.update({
          where: { id: isValidCartItem.product.id },
          data: {
            stockLevel: {
              increment: quantity,
            },
          },
        })

        // remove item from cart
        const updatedCart = prisma.cart.update({
          where: { id: cartId },
          data: {
            cartItems: { disconnect: { id: cartItemId } },
          },
        })

        // wrap operations in transaction to ensure both mutations succeed
        const result = await prisma.$transaction([newStockLevel, updatedCart])

        // if both mutations succeed, delete cart item
        if (result.every((item) => item.id)) {
          await prisma.cartItem.delete({
            where: { id: cartItemId },
          })
        }
        return isValidCart
      } catch (error) {
        formatError('removeFromCart', error)
        return error
      }
    },
    checkout: async (
      _parent: unknown,
      { cartId }: MutationCheckoutArgs,
      { prisma }: Context,
    ) => {
      try {
        const cart = await prisma.cart.findFirst({
          where: { AND: [{ id: cartId }, { isCheckedOut: false }] },
        })

        if (!cart) {
          throw new ApolloError('Cart not found')
        }

        // checkout cart
        const updatedCart = await prisma.cart.update({
          where: { id: cart.id },
          data: {
            isCheckedOut: true,
          },
        })

        return updatedCart
      } catch (error) {
        formatError('checkout', error)
        return error
      }
    },
  },
  Cart: {
    // format price into rands so we don't have to worry about that on the frontend
    cartItems: async (_parent: Cart, args: unknown, { prisma }: Context) => {
      try {
        // get  cart items in cart
        const cartItems = await prisma.cart
          .findUnique({
            where: { id: _parent.id },
          })
          .cartItems()

        return cartItems
      } catch (error) {
        formatError('Cart.cartItems', error)
        return error
      }
    },
    // calculate price based db so we don't have to worry about outdated data
    total: async (_parent: Cart, args: unknown, { prisma }: Context) => {
      try {
        // get all cart items
        const cartItems = await prisma.cart
          .findUnique({
            where: { id: _parent.id },
          })
          .cartItems()

        // sum up all cart items prices for total
        const total = cartItems.reduce((acc, item) => acc + item.price, 0)

        return formatPrice(total)
      } catch (error) {
        formatError('Cart.total', error)
        return error
      }
    },
  },
}
