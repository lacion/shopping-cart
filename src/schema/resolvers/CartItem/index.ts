import { CartItem } from '@prisma/client'
import { ApolloError } from 'apollo-server'
import { MutationUpdateQuantityArgs } from '../../../generated'
import { Context, formatError, formatPrice } from '../../../utils'

export default {
  Mutation: {
    updateQuantity: async (
      _parent: unknown,
      { input }: MutationUpdateQuantityArgs,
      { prisma }: Context,
    ) => {
      try {
        const { quantity, cartItemId } = input

        // check for valid quantity
        if (quantity < 1) {
          throw new ApolloError('Invalid quantity')
        }

        // no specific item id passed.
        if (!cartItemId) {
          throw new ApolloError('Item not found')
        }

        // check if product is in stock based on quantity
        const hasEnoughProducts = await prisma.product.findFirst({
          where: {
            AND: [
              { cartItems: { some: { id: cartItemId } } },
              { stockLevel: { gte: quantity } },
            ],
          },
        })

        // if not enough products, throw error
        if (!hasEnoughProducts) {
          throw new ApolloError('Product out of stock')
        }

        // calculate item price
        const price = hasEnoughProducts.price * quantity

        // add product to cart
        const newCartItem = prisma.cartItem.update({
          where: { id: cartItemId },
          data: {
            quantity,
            price,
          },
        })

        // update product stock level
        const newStockLevel = prisma.product.update({
          where: { id: hasEnoughProducts.id },
          data: {
            stockLevel: {
              decrement: quantity,
            },
          },
        })

        // wrap operations in transaction to ensure both mutations succeed
        await prisma.$transaction([newCartItem, newStockLevel])

        return newCartItem
      } catch (error) {
        formatError('updateQuantity', error)
        return error
      }
    },
  },
  CartItem: {
    // format price into rands so we don't have to worry about that on the frontend
    price: async (_parent: CartItem, args: unknown, { prisma }: Context) => {
      try {
        // get current cart item
        const cartItem = await prisma.cartItem.findUnique({
          where: { id: _parent.id },
        })

        return formatPrice(cartItem.price)
      } catch (error) {
        formatError('Cart.total', error)
        return error
      }
    },
  },
}
