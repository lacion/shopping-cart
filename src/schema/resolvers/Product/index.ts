import { Prisma, ProductWhereInput } from '@prisma/client'
import { ApolloError } from 'apollo-server'
import { QueryProductsArgs } from '../../../generated'
import { Context, formatError } from '../../../utils'

export default {
  Query: {
    products: async (
      _parent: unknown,
      { filterBy }: QueryProductsArgs,
      { prisma }: Context,
    ) => {
      try {
        const { categoryId, name, skip, take } = filterBy
        let and: Prisma.Enumerable<ProductWhereInput> = []

        // no specific filter passed.
        if (!categoryId && !name) {
          return prisma.product.findMany({ skip, take })
        }

        // filter by category
        if (categoryId) {
          // check if categoryId is valid
          const category = await prisma.category.findUnique({
            where: { id: categoryId },
          })

          if (!category) {
            throw new ApolloError('Category not found')
          }

          and.push({ category: { id: categoryId } })
        }

        // search by product name
        if (name) {
          and.push({ name: { contains: name } })
        }

        return prisma.product.findMany({
          skip,
          take,
          where: {
            AND: and,
          },
        })
      } catch (error) {
        formatError('products', error)
        return error
      }
    },
  },
}
