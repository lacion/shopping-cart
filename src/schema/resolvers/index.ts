import merge from 'lodash/merge'
import Auth from './Auth'
import Cart from './Cart'
import CartItem from './CartItem'
import Product from './Product'

export default merge({}, Auth, Cart, CartItem, Product)
