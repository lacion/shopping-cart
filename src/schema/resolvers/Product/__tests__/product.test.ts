import { Context, createMockContext, MockContext } from '../../../../utils'
import resolvers from '../index'

let mockCtx: MockContext
let context: Context

beforeEach(() => {
  mockCtx = createMockContext()
  context = mockCtx as unknown as Context
})

describe('Product - get products', () => {
  it('returns an error if invalid category id is parsed', async () => {
    // setup
    const {
      Query: { products },
    } = resolvers

    let args = { filterBy: { categoryId: 'thisDoesNotExist' } }

    // test
    const result = await products({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })

  it('returns a response of type array', async () => {
    // setup
    const {
      Query: { products },
    } = resolvers

    let args = { filterBy: { skip: 0, take: 10 } }

    const product = {
      id: '1',
      name: 'my product',
      price: 2300,
      sku: 99888,
      stockLevel: 45,
      expiresAt: new Date(),
      categoryId: 'string',
    }

    mockCtx.prisma.product.findMany.mockResolvedValue([product])

    // test
    const result = await products({}, args, context)

    // assert
    expect(Array.isArray(result)).toBe(true)
  })
})
