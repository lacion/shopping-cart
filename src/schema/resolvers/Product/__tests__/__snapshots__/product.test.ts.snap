// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product - get products returns an error if invalid category id is parsed 1`] = `[Error: Category not found]`;
